### PR TITLE
Server: add with_shared_state()

### PR DESCRIPTION
**Pull request by Fishrock123**
Allows true state sharing for multiple tide server applications without `Arc<Arc<State>>`.

See discord discussion: https://discordapp.com/channels/598880689856970762/649056551835009097/731205374832541746

### CODE REVIEW

**Comment by yoshuawuyts** [comment]
I can't open the discord link so I'm not sure what the motivation is — but guessing it seems like we want to have access to `Arc<State>` somehow?

Not opposed to always have require State to be `Arc`d when passed to the `with_state` constructor — or perhaps it's even better to require `State: Clone` which would provide the same guarantees in a more direct way.

**Comment by Fishrock123** [comment]
Yeah, the idea is to be able to have more than one server share the same state without passing an `Arc`, which would become `Arc<Arc<>>` and then passed to middleware as an `Arc<State>`.

I think this was the intention to begin with but probably was mistaken?

Does `State: Clone` prevent us from having an `Arc` internally that is different from a user-passed `Arc`?

**Comment by yoshuawuyts** [comment]
`State: Clone` indeed means we would no longer need to have an internal Arc since the state can be freely cloned already. `with_state(Arc::new(state))` would be valid.

One downside however is that depending on how State is constructed, `Arc` may show up in the Server's type signature — but I don't think this needs to be a problem.

**Comment by Fishrock123** [comment]
I think that was part of the stated issue though, that it's annoying to have to deal with an `Arc<>` in the middleware when it's probably going to be anyways?

**Comment by yoshuawuyts** [comment]
I mean in some cases it may very well be an `Arc<RwLock>` or perhaps only some fields need to be wrapped in an `Arc<Mutex>` for the whole struct to simply be `#[derive (Clone)]`. There are patterns we can introduce people to that extend to Rust as a whole.

I think `State: Clone` is probably the best way ahead here.

**Comment by Fishrock123** [comment]
Uhhh, I'm not sure that is possible. `impl<State> Read for Request<State>` requires `DerefMut`...

**Comment by Fishrock123** [comment]
So @jbr got the above to compile using `pin-project-lite`: https://gist.github.com/jbr/b0fd2964eb57a1d8d512e4d167ad6554

However I think that `Request` probably shouldn't hold the `State`. See https://github.com/http-rs/tide/issues/643

File: src/lib.rs [hunk 1]
```diff
@@ -283,5 +285,47 @@ where
...
+/// #
+/// # Ok(()) }) }
+/// ```
+pub fn with_shared_state<State>(state: Arc<State>) -> server::Server<State>
+where
+    State: Send + Sync + 'static,
+{
+    Server::with_shared_state(state)
+}
+
```
Comment (yoshuawuyts) [inline src/lib.rs:329]: Now that `State: Clone` this method is no longer needed.

```suggestion
```

File: src/server.rs [hunk 2]
```diff
@@ -167,7 +167,7 @@ impl Default for Server<()> {
 }
 
 impl<State: Send + Sync + 'static> Server<State> {
-    /// Create a new Tide server with shared application scoped state.
+    /// Create a new Tide server with application scoped state.
```
Comment (yoshuawuyts) [inline src/server.rs:170]: Application state is and always was shared within the application. Now that `State: Clone` it's true both locally to applications, and between applications.

```suggestion
    /// Create a new Tide server shared with application scoped state.
```

**Comment by Fishrock123** [comment]
Closing in favor of https://github.com/http-rs/tide/pull/644

**Pull request closed without merging**